### PR TITLE
Base class for scripts

### PR DESCRIPTION
**Pull request by kbwestfall**
This PR introduces a new base class for all the scripts: `pypeit.scripts.scriptbase.ScriptBase`

There were two goals for this:
 - Streamline the script construction
 - Fix the fact that doc updates basically always lead to new doc/help/*.rst files.

Because all scripts now use `pypeit.scripts.scriptbase.ScriptBase` as the base class, this allows us to automatically find and document all the scripts, similar to how this is done for the `Spectrograph` classes.

The procedure for creating a new script is now:
 - Write the script `get_parser` and `main` functions in basically the same way that you've done before, within a class that inherits from `pypeit.scripts.scriptbase.ScriptBase`
 - Add the module to the `scripts/__init__.py` file; i.e., `from pypeit.scripts import new_script`
 - Add the relevant line to the `setup.cfg` file; i.e., under `console_scripts` add something like `pypeit_new_script = pypeit.scripts.new_script:NewScript.entry_point`

Note that the script files in the `pypeit/scripts` directory:
 - Should not be executable (i.e., no `x`s in their permissions)
 - Should not start with a env like, `#!/usr/bin/env python`
 - Should not end with the `if __name__ == '__main__':` block

Creating the executables from the raw `*.py` script files is all handled by the pip install.

It would be nice if the entries in `setup.cfg` could somehow be automated given that all the elements of these lines can be accessed programmatically now.  See `pypeit.scripts.__init__.script_classes`.  Any thoughts @tepickering ?

In the process of going through this, I deprecated what I thought were out-of-date scripts.  Namely, `arcid_plot`, `chk_tilts`, `find_objects`, and `lowrdx_pixflat`.  If any of these should not be deprecated, it's easy enough to put them back.  Just let me know.

Although many files are touched, the new code is pretty minimal, and largely confined to `pypeit.scripts.scriptbase` and `pypeit.scripts.__init__`.  All tests pass (I fixed the one that is skipped PR #1198, which should be merged before this):
```
==== 280 passed, 1 skipped in 937.60s (0:15:37) ====
```
I haven't run the dev-suite, but this should be straight-forward.

### CODE REVIEW

**Comment by profxj** [comment]
@rcooke-ast -- Please comment on the list of deprecated scripts.

**Review by profxj — APPROVED** [review]
Damn, nice refactor!

Main request from me:  Can you add the details you put at the
top of this PR into the `scripts.rst` file?  In a `Developer` section.

Thanks!

File: pypeit/scripts/__init__.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
+
+from pypeit.utils import all_subclasses
+from pypeit.scripts import scriptbase
+
+# The import of all the script modules here is what enables the dynamic
+# compiling of all the available scripts below
```
Comment (tepickering) [inline pypeit/scripts/__init__.py:6]: i wonder if there's a clever way to use `import *` here to make this more succinct and automatically include new scripts as they get added.

**Comment by tepickering** [comment]
made a quick pass and it looks 🤘 great! 🤘 there's a lot going on, though, and the github diffs make some of it hard to follow. i will check this out locally to look at more closely tomorrow and do a proper review.

**Comment by kbwestfall** [comment]
Updates:
 - Turns out I found something that python 3.9 can do that python 3.8, which was causing the testing workflow to fault.  I've fixed that, and we can update when/if we start to require 3.9.
 - I've added a `new_script` doc that is pointed to at the top of the `scripts` doc and linked to in the dev doc.
 - I've confirmed `pypeit_obslog` is documented.  Most of it is in the `setup` doc, but I've now added references to this in the `cookbook` and `scripts` docs.
 - I've merged this with the `release` branch, meaning the `develop` branch will be synced with it when this is merged.

Can someone with a beefy computer run the dev-suite on this branch?

File: pypeit/scripts/__init__.py (hunk 1 above)
Comment (kbwestfall) [reply pypeit/scripts/__init__.py:6]: Yeah, I had the same thought.  It may be an outdated concern, but I've been warned off using `from whatever import *`.  If it is done, you're supposed to put a `__all__` list in the relevant `__init__` file, which means you need to list all of the relevant modules anyway.  I'm certainly open to suggestions, but I couldn't think of a clever way to do this.  If we do find a clever way, we should put it in `spectrographs/__init__.py`, as well.

**Comment by kbwestfall** [comment]
Thanks for starting the review, @tepickering!  Yes, unfortunately, the diff in github isn't all that helpful.  The combination of the class indentation and my OCD limiting the code to 100 characters per line, means that most of the script files are logged as changed even when the changes are pretty minimal.

**Comment by profxj** [comment]
I'll run the Dev Suite tomorrow

File: pypeit/scripts/__init__.py (hunk 1 above)
Comment (tepickering) [reply pypeit/scripts/__init__.py:6]: places like `__init__.py` are cases where it is ~ok to use it. `photutils` is one codebase that i am familiar with that does, but does put `#noqa` on those lines to keep linters from whinging about it.

**Comment by profxj** [comment]
All DevSuite tests pass:

--- PYPEIT DEVELOPMENT SUITE PASSED 115/115 TESTS (Masters ignored) ---

**Comment by tepickering** [comment]
it looks like we can at least use the `file:` directive in `setup.cfg` to load its `entry_points` config: https://setuptools.readthedocs.io/en/latest/userguide/declarative_config.html. might also be worth looking at the capabilities for dynamic discovery that `importlib.metadata` provides: https://setuptools.readthedocs.io/en/latest/userguide/entry_point.html#dynamic-discovery-of-services-and-plugins. it looks like that might be a cleaner way of doing what you do in `__init__.py`. it would at least enable all the `entry_point` configuration to be defined in one place, `setup.cfg` or a file it loads.

**Comment by kbwestfall** [comment]
> it looks like we can at least use the `file:` directive in `setup.cfg` to load its `entry_points` config: https://setuptools.readthedocs.io/en/latest/userguide/declarative_config.html. might also be worth looking at the capabilities for dynamic discovery that `importlib.metadata` provides: https://setuptools.readthedocs.io/en/latest/userguide/entry_point.html#dynamic-discovery-of-services-and-plugins. it looks like that might be a cleaner way of doing what you do in `__init__.py`. it would at least enable all the `entry_point` configuration to be defined in one place, `setup.cfg` or a file it loads.

Thanks for looking into this, @tepickering , but can we punt on this for now?  With the tests and dev-suite passed, I'm wary of editing the code again in a way that would require re-running the dev-suite.

**Comment by tepickering** [comment]
oh yeah, for sure. first make it work, then make it cool. i also want to research it further and play around some more to get a better handle on how best to optimize this.

**Review by tepickering — APPROVED** [review]
i really appreciate this clean refactoring that does away with the copy pasta that i introduced a while back. since it's passing all of the tests and the dev suite, i consider it gtg to merge. 👍
